### PR TITLE
CONDEC-672: Improve classifier performance

### PR DESCRIPTION
**Pull request by desombre**
Two main focuses are tackled.
Firstly the logic for classifying updated issue comments needs too be reworked. Secondly the default training file will be balanced and reduced.

### CODE REVIEW

**Comment by kleebaum** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestPreprocessorImpl.java  1
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestFileTrainer.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760&bid=16091328)

File: src/main/java/de/uhd/ifi/se/decision/management/jira/classification/preprocessing/PreprocessorImpl.java [hunk 1]
```diff
@@ -163,7 +163,7 @@ public synchronized List preprocess(String sentence) {
 		}catch (Exception e){
 			System.out.println(e.getMessage());
 			initFiles();
-			throw e;
+			throw new Exception(e);
```
Comment (kleebaum) [inline src/main/java/de/uhd/ifi/se/decision/management/jira/classification/preprocessing/PreprocessorImpl.java]: Codacy Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760)

File: src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestOnlineTrainer.java [hunk 2]
```diff
@@ -7,16 +7,13 @@
...
 import org.junit.Test;
 
 import java.io.*;
 import java.util.ArrayList;
 import java.util.List;
 
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.*;
```
Comment (kleebaum) [inline src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestOnlineTrainer.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit.Assert'](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760)

**Comment by kleebaum** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestPreprocessorImpl.java  1
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestFileTrainer.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760&bid=16091328)

File: src/main/java/de/uhd/ifi/se/decision/management/jira/classification/preprocessing/PreprocessorImpl.java (hunk 1 above)
Comment (kleebaum) [inline src/main/java/de/uhd/ifi/se/decision/management/jira/classification/preprocessing/PreprocessorImpl.java]: Codacy Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760)

File: src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestOnlineTrainer.java [hunk 3]
```diff
@@ -7,16 +7,13 @@
...
```
Comment (kleebaum) [inline src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestOnlineTrainer.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit.Assert'](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760)

**Comment by kleebaum** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestPreprocessorImpl.java  1
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestFileTrainer.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760&bid=16091328)

File: src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestOnlineTrainer.java (hunk 3 above)
Comment (kleebaum) [inline src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestOnlineTrainer.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit.Assert'](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760)

File: src/main/java/de/uhd/ifi/se/decision/management/jira/classification/preprocessing/PreprocessorImpl.java (hunk 1 above)
Comment (kleebaum) [inline src/main/java/de/uhd/ifi/se/decision/management/jira/classification/preprocessing/PreprocessorImpl.java]: Codacy Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760)

**Comment by kleebaum** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestFileTrainer.java  2
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestPreprocessorImpl.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760&bid=16091328)

File: src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestOnlineTrainer.java (hunk 3 above)
Comment (kleebaum) [inline src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestOnlineTrainer.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit.Assert'](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760)

File: src/main/java/de/uhd/ifi/se/decision/management/jira/classification/preprocessing/PreprocessorImpl.java (hunk 1 above)
Comment (kleebaum) [inline src/main/java/de/uhd/ifi/se/decision/management/jira/classification/preprocessing/PreprocessorImpl.java]: Codacy Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760)

**Comment by kleebaum** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestPreprocessorImpl.java  1
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestFileTrainer.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760&bid=16091328)

File: src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestOnlineTrainer.java (hunk 3 above)
Comment (kleebaum) [inline src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestOnlineTrainer.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit.Assert'](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760)

File: src/main/java/de/uhd/ifi/se/decision/management/jira/classification/preprocessing/PreprocessorImpl.java (hunk 1 above)
Comment (kleebaum) [inline src/main/java/de/uhd/ifi/se/decision/management/jira/classification/preprocessing/PreprocessorImpl.java]: Codacy Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760)

File: src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestFileTrainer.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import net.java.ao.test.jdbc.NonTransactional;
+import org.junit.Test;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+
+import static junit.framework.TestCase.*;
```
Comment (kleebaum) [inline src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestFileTrainer.java]: Codacy Issue found: [Avoid unused imports such as 'junit.framework.TestCase'](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760)

**Comment by kleebaum** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestPreprocessorImpl.java  1
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestFileTrainer.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760&bid=16091328)

File: src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestFileTrainer.java (hunk 4 above)
Comment (kleebaum) [inline src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestFileTrainer.java]: Codacy Issue found: [Avoid unused imports such as 'junit.framework.TestCase'](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760)

File: src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestOnlineTrainer.java (hunk 3 above)
Comment (kleebaum) [inline src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestOnlineTrainer.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit.Assert'](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760)

File: src/main/java/de/uhd/ifi/se/decision/management/jira/classification/preprocessing/PreprocessorImpl.java (hunk 1 above)
Comment (kleebaum) [inline src/main/java/de/uhd/ifi/se/decision/management/jira/classification/preprocessing/PreprocessorImpl.java]: Codacy Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760)

**Comment by kleebaum** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestFileTrainer.java  2
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestPreprocessorImpl.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760&bid=16091328)

**Comment by kleebaum** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestFileTrainer.java  2
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestPreprocessorImpl.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760&bid=16091328)

**Comment by kleebaum** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestPreprocessorImpl.java  1
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestFileTrainer.java  2
- src/main/java/de/uhd/ifi/se/decision/management/jira/eventlistener/JiraIssueTextExtractionEventListener.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760&bid=16091328)

File: src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/versioncontrol/CommitMessageToCommentTranscriber.java [hunk 5]
```diff
@@ -83,26 +80,37 @@ public void postComment(ApplicationUser user, RevCommit commit) {
...
 		if (Boolean.parseBoolean(ConfigPersistenceManager.getValue(projectKey, "isPostFeatureBranchCommitsActivated"))) {
+
+			Optional.ofNullable(gitClient.getFeatureBranchCommits(branch)).ifPresent(featureBranchCommits::addAll);
 			for (RevCommit commit : this.featureBranchCommits) {
-				this.postComment(defaultUser, commit);
+				this.postComment(defaultUser, commit, branch);
 			}
 		}
+		if (branch.getName().contains("develop") || branch.getName().contains("master")) {
+			if (Boolean.parseBoolean(ConfigPersistenceManager.getValue(projectKey, "isPostSquashedCommitsActivated"))) {
```
Comment (kleebaum) [inline src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/versioncontrol/CommitMessageToCommentTranscriber.java]: Codacy Issue found: [These nested if statements could be combined](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760)

**Comment by kleebaum** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestFileTrainer.java  2
- src/main/java/de/uhd/ifi/se/decision/management/jira/eventlistener/JiraIssueTextExtractionEventListener.java  1
- src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/versioncontrol/CommitMessageToCommentTranscriber.java  1
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestPreprocessorImpl.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760&bid=16091328)

File: src/main/java/de/uhd/ifi/se/decision/management/jira/rest/impl/ViewRestImpl.java [hunk 6]
```diff
@@ -47,6 +24,22 @@
 import de.uhd.ifi.se.decision.management.jira.view.treeviewer.TreeViewer;
 import de.uhd.ifi.se.decision.management.jira.view.vis.VisGraph;
 import de.uhd.ifi.se.decision.management.jira.view.vis.VisTimeLine;
+import org.eclipse.jgit.lib.Ref;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.*;
```
Comment (kleebaum) [inline src/main/java/de/uhd/ifi/se/decision/management/jira/rest/impl/ViewRestImpl.java]: Codacy Issue found: [Avoid unused imports such as 'javax.ws.rs'](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760)

**Comment by kleebaum** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestFileTrainer.java  2
- src/test/java/de/uhd/ifi/se/decision/management/jira/classification/TestPreprocessorImpl.java  1
- src/main/java/de/uhd/ifi/se/decision/management/jira/extraction/versioncontrol/CommitMessageToCommentTranscriber.java  2
- src/main/java/de/uhd/ifi/se/decision/management/jira/eventlistener/JiraIssueTextExtractionEventListener.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760&bid=16091328)

File: src/main/java/de/uhd/ifi/se/decision/management/jira/rest/impl/ViewRestImpl.java (hunk 6 above)
Comment (kleebaum) [inline src/main/java/de/uhd/ifi/se/decision/management/jira/rest/impl/ViewRestImpl.java]: Codacy Issue found: [Avoid unused imports such as 'javax.ws.rs'](https://app.codacy.com/manual/UHD/cures-condec-jira/pullRequest?prid=4848760)

**Comment by desombre** [comment]
An error in the filtering of comments that were to be classified was fixed. All comments were filtered out before classification.
The bug that no squashed merge commits were shown should now be fixed. When the branch was deleted from which the merge was conducted, the logic  for posting commit messages was skipped altogether. This was fixed by checking "master" and "develop" branches for those squashed commits. It would propably be better if these branch names were configurable.

**Pull request merged**